packages/toolpad-app/src/server/auth.ts

```diff
@@ -1,64 +1,127 @@
 import express, { Router } from 'express';
 import { Auth } from '@auth/core';
-import GithubProvider from '@auth/core/providers/github';
+import GithubProvider, { GitHubEmail, GitHubProfile } from '@auth/core/providers/github';
 import GoogleProvider from '@auth/core/providers/google';
 import { getToken } from '@auth/core/jwt';
+import { AuthConfig, TokenSet } from '@auth/core/types';
+import { OAuthConfig } from '@auth/core/providers';
 import { asyncHandler } from '../utils/express';
 import { adaptRequestFromExpressToFetch } from './httpApiAdapters';
 import { ToolpadProject } from './localMode';
 import * as appDom from '../appDom';
 
-export function createAuthHandler(base: string): Router {
+const githubProvider = GithubProvider({
+  clientId: process.env.TOOLPAD_GITHUB_ID,
+  clientSecret: process.env.TOOLPAD_GITHUB_SECRET,
+  userinfo: {
+    url: 'https://api.github.com/user',
+    async request({
+      tokens,
+      provider,
+    }: {
+      tokens: TokenSet;
+      provider: OAuthConfig<GitHubProfile>;
+    }) {
+      const headers = {
+        Authorization: `Bearer ${tokens.access_token}`,
+        'User-Agent': 'authjs',
+      };
+
+      const profile = await fetch(provider.userinfo?.url as URL, {
+        headers,
+      }).then(async (githubRes) => githubRes.json());
+
+      if (!profile.email) {
+        // If the user does not have a public email, get another via the GitHub API
+        // See https://docs.github.com/en/rest/users/emails#list-public-email-addresses-for-the-authenticated-user
+        const githubRes = await fetch('https://api.github.com/user/emails', {
+          headers,
+        });
+
+        if (githubRes.ok) {
+          const emails: GitHubEmail[] = await githubRes.json();
+          profile.email = (emails.find((e) => e.primary) ?? emails[0]).email;
```

Comment:
Can't we just look for the first verified email that satisfies our restrictedDomains setting and use that as the email? Then set the `email_verified` boolean to `true`?

Reply:
That was the default logic but it would be an improvement, i'll change it and fallback to primary email.

Reply:
I've added the changes here https://github.com/mui/mui-toolpad/pull/3047/commits/fd18d7c074bc53f1e826e61235a34089af2237ea, let me know if it's good now!